src/AI/Muriel.java

```diff
@@ -0,0 +1,750 @@
+    log(String.format("  Action: %s", action));
+    return action;
+  }
+
+  private boolean shouldAttack(Unit unit, Unit target, GameMap gameMap)
+  {
+    // Calculate the cost of the damage we can do.
+    BattleInstance.BattleParams params = new BattleInstance.BattleParams(unit, unit.chooseWeapon(target.model, 1, true), target, gameMap.getLocation(target.x, target.y).getEnvironment());
+    double damage = params.calculateDamage();
+    UnitMatchupAndMetaInfo umami = myUnitEffectMap.get(new UnitModelPair(unit.model, target.model));
+
+    // This attack is a good idea if our cost effectiveness is in the acceptable range, or if we can at least half-kill them.
+    // The second check is needed because one glass cannon may not have a great overall ratio against another; whoever hits first wins, e.g. Mech vs Anti-Air.
+    return (umami.costEffectivenessRatio > COST_EFFECTIVENESS_THRESHOLD) || (damage > (target.getHP() / 2.0));
+  }
+
+  private void queueUnitProductionActions(GameMap gameMap)
+  {
+    log("Evaluating Production needs");
+    int budget = myCo.money;
+
+    // Get a count of enemy forces.
+    Map<UnitModel, Double> enemyUnitCounts = new HashMap<UnitModel, Double>();
+    for( Commander co : enemyCos )
+    {
+      for( Unit u : co.units )
+      {
+        if( !u.model.hasDirectFireWeapon() ) continue; // Only handle direct-fire units for now.
+        // Count how many of each model of enemy units are in play.
+        if( enemyUnitCounts.containsKey(u.model))
+        {
+          enemyUnitCounts.put(u.model, enemyUnitCounts.get(u.model) + (u.getHP() / 10) );
+        }
+        else
+        {
+          enemyUnitCounts.put(u.model, u.getHP() / 10.0 );
+        }
+      }
+    }
+
+    // Figure out what unit types we can purchase with our available properties.
+    CommanderProductionInfo CPI = new CommanderProductionInfo(myCo, gameMap);
+
+    if( CPI.availableProperties.isEmpty() )
+    {
+      log("No properties available to build.");
+      return;
+    }
+
+    // Order the enemy units by model.cost (not number), lowest first.
+    ArrayList<UnitModel> enemyModels = new ArrayList<UnitModel>();
+
+    // Sort enemy units by cardinality. We will attempt to build counters for the most numerous first.
+    ArrayList<Entry<UnitModel, Double>> entryArray = new ArrayList<Entry<UnitModel, Double>>(enemyUnitCounts.entrySet());
+    Collections.sort(entryArray, new UnitQuantityComparator());
+    for( Entry<UnitModel, Double> ent : entryArray )
+    {
+      enemyModels.add(ent.getKey());
+    }
+
+    // Try to purchase units that will counter the most-represented enemies.
+    ArrayList<PurchaseOrder> shoppingCart = new ArrayList<PurchaseOrder>();
+    while( !enemyModels.isEmpty() && !CPI.availableUnitModels.isEmpty())
+    {
+      // Find the first (most numerous) enemy UnitModel, and remove it. Even if we can't find an adequate counter,
+      // there is not reason to consider it again on the next iteration.
```

Comment:
This doesn't work because the UnitQuantityComparator sorts in *a*scending order, not descending order.

Reply:
In practice, ascending order seems more robust. Really, I want to rework how it does the production stuff, but that'll have to be a future patch.

Reply:
How is ascending order more robust in practice? (I suppose I can see not wanting to try to counter infantry each turn, heh.)
At the least, your documentation should be updated.

Reply:
Yes. Done.